rfcs/0082-lib-experimental.md

```diff
@@ -0,0 +1,129 @@
+---
+feature: lib.experimental
+start-date: 2020-12-17
+author: Francois-Rene Rideau
+co-authors: N/A
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+I propose that we create a directory lib/experimental for experimental code,
+also available as lib.experimental and pkgs.lib.experimental. Such code is
+available for limited use within select parts of nixpkgs, with the explicit
+opt-in approval of their respective maintainers, but not for widespread use
+in other parts of nixpkgs.
+
+# Motivation
+[motivation]: #motivation
+
+Some libraries we *must* use
+mandatorily unless it's impossible or at least unreasonable not to.
+Some libraries we *should* use
+unless there's a good reason not to.
+Some libraries we *may* use
+according to each contributors' preferences.
+Some libraries we *may not* use
+without an explicit endorsement by enough maintainers of the relevant packages.
+Some libraries we *shouldn't* use
+without a strong reason to do it.
+Some libraries we *mustn't* use
+and is only left temporarily for backward compatibility until complete migration off of it.
+
+I propose we create a directory `lib/experimental`
+to tag libraries as explicitly being part of the *may not* category.
+
+Creating this experimental section to nixpkgs makes it clearer to maintainers
+which code they may use where, and which code they may not.
+If you are the maintainer of a package, and understand the risks related to using
+experimental code, you may use this code or accept patches using it.
+But you may also freely reject it, and no one should impose you that code without your consent.
+In particular you understand that the code and its API may change, and
+you may have to update your code accordingly, or accept patches that do.
+Thus you agree to collaborate with the author of the experimental code
+while the code remains in the experimental section.
+
+Code that should *not* go into `lib.experimental` but directly into `lib` include:
+* code that is good quality, well tested, known to work, with a stable API,
+  good documentation, and no controversy among maintainers of nixpkgs.
+
+Code that should *not* go into `lib.experimental` and not into nixpkgs at all include:
+* code that bad quality, untested, not working, etc.
+* code that is not yet used by any package in nixpkgs nor by any extension to nixpkgs.
+* code that while cool has no purpose in nixpkgs as such and can well live outside it.
+
+Making these social constraints explicit avoid accidents and conflict that would happen
+if they were left implicit.
```

Comment:
What is the cost of these accidents and conflict. Could you provide real-world examples?

Also note that small accidents and conflicts (human or otherwise) are part of how some software evolves.

Reply:
Not sure about real-world examples in nixpkgs... maybe this is all overkill, but I was told not to make my pop.nix library part of the regular `lib/` for now, so I wanted a way to get it in without it being "regular".